pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh

```diff
@@ -653,18 +653,45 @@ fi
 # If we're not just building, then make the new configuration the boot
 # default and/or activate it now.
 if [[ "$action" = switch || "$action" = boot || "$action" = test || "$action" = dry-activate ]]; then
+    # Using systemd-run here to protect against PTY failures/network
+    # disconnections during rebuild.
+    # See: https://github.com/NixOS/nixpkgs/issues/39118
+    cmd=(
+        "systemd-run"
+        "-E" "LOCALE_ARCHIVE" # Will be set to new value early in switch-to-configuration script, but interpreter starts out with old value
+        "--collect"
+        "--no-ask-password"
+        "--pty"
+        "--quiet"
+        "--same-dir"
+        "--service-type=exec"
+        "--unit=nixos-rebuild-switch-to-configuration"
+        "--wait"
+    )
+    # Check if we have a working systemd-run. In chroot environments we may have
+    # a non-working systemd, so we fallback to not using systemd-run.
+    # You may also want to explicitly set NIXOS_IGNORE_SYSTEMD_RUN environment
+    # variable, since systemd-run runs inside an isolated environment and
+    # this may break some post-switch scripts. However keep in mind that this
+    # may be dangerous in remote access (e.g. SSH).
+    if [[ -z "$NIXOS_IGNORE_SYSTEMD_RUN" ]] && targetHostCmd "${cmd[@]}" true &>/dev/null; then
+        logVerbose "Using systemd-run to switch the configuration."
+    else
+        logVerbose "Skipping systemd-run to switch configuration since it is not working in target host or NIXOS_IGNORE_SYSTEMD_RUN env is set."
```

Comment:
I think this should be a warning, because this option is not sustainable in the long run.
We can burden neither users nor maintainers with this distinction, so it should only exist as an escape hatch for a limited time. (A year?)
Scripts that relied on this should be fixed.

Reply:
Done.

Reply:
@roberth What do you mean? We *can't* use `systemd-run` in a chroot, so this has to stay if we want to be able to `nixos-rebuild` in a `nixos-enter`.

Reply:
@ElvishJerricco you're right!

The warning should only apply when `NIXOS_IGNORE_SYSTEMD_RUN` is set.

Reply:
Fixed @roberth @ElvishJerricco.